src/rgw/rgw_sal_motr.cc

```diff
@@ -3493,7 +3493,30 @@ int MotrStore::list_users(const DoutPrefixProvider* dpp, const std::string& meta
                         std::string& marker, int max_entries, void *&handle,
                         bool* truncated, std::list<std::string>& users)
 {
-    return 0;
+  int rc;
+  bool max_entires_specified = true;
+  if(max_entries <= 0) {
+    max_entries = 1000; 
+    max_entires_specified = false;
```

Comment:
1. Upperbound of 1000 is not applied.
2. max_entries_specified boolean is not needed. 
    * The keys are truncated if we are able to fetch an extra key than max_entries, else it is not. 
    * Even if user doesn't specified a limit but we have say 10k users, still we want to truncate the user list to 1k.

Reply:
> max_entries_specified boolean is not needed.
The keys are truncated if we are able to fetch an extra key than max_entries, else it is not.
Even if user doesn't specified a limit but we have say 10k users, still we want to truncate the user list to 1k.

How about user does not specify max entries and we have less than thousand users, in that case truncated should be false but by aove logic it will be true. And thats why `max_entries_specified` is required.

Reply:
> Upperbound of 1000 is not applied.

Sorry I didn't understand this, can you elaborate further?

Reply:
> Sorry I didn't understand this, can you elaborate further?

One question to confirm if what I am thinking applies here, If we have 2000 keys and we don't provide a max-entries parameter then should it show all 2000 keys or 1000 keys with a next-marker key?

Reply:
> One question to confirm if what I am thinking applies here, If we have 2000 keys and we don't provide a max-entries parameter then should it show all 2000 keys or 1000 keys with a next-marker key?

It should show 1000 keys with a next-marker. In any case it should not print more than 1000 keys.

Reply:
In that case, we need a check `max_entries > 1000` as well to truncate the result.
<details> <summary>This should work </summary>

```c++
{
  int rc;
  if (max_entries <= 0 or max_entries > 1000)
    max_entries = 1000; 
    
  vector<string> keys(max_entries + 1);
  vector<bufferlist> vals(max_entries + 1);

  keys[0] = marker;
  rc = next_query_by_name(RGW_MOTR_USERS_IDX_NAME, keys, vals);
  if (rc < 0) {
    ldpp_dout(dpp, 0) << "ERROR: NEXT query failed. " << rc << dendl;
    return rc;
  }
  if (!(keys.back()).empty()) {
    *truncated = true;
    marker = keys.back(); 
  }
  for (int i = 0; i < int(keys.size()) - 1; i++) {
    if (keys[i].empty())
      break;
    users.push_back(keys[i]);
  }
  return rc;
}
```

</details>

Also, we are fetching metadata of every user but only keys are needed.
so If we have some motrlib function to fetch only the keys then that might save some time/memory. 
@andriytk, @siningwuseagate ??

Reply:
During development I had a test case in mind which I can't think of right now. So removing `max_entries_specified`

Reply:
@andriytk I don't see `M0_IC_LIST` case in `do_idx_op()`. 
Is there any specific attribute required for it and what should be the key here?

Reply:
> During development I had a test case in mind which I can't think of right now. So removing max_entries_specified

👍  Thanks Jeet!!

> @andriytk I don't see M0_IC_LIST case in do_idx_op().
Is there any specific attribute required for it and what should be the key here?

Is there documentation with all motrlib function/api definitions to refer etc.? @andriytk, @DPG17 ?

Reply:
Found this from cortx-motr repo @jjxsg, pls check if it is useful
```c
rc = m0_idx_op(&idx, M0_IC_LIST, &keys, NULL, rcs, flags, op);
```

Reply:
Yes, it is an example of its usage: https://github.com/Seagate/cortx-motr/blob/2d8a769326a8c07f4827ffee97615946895f6ad7/motr/m0kv/index_op.c#L139-L140

Reply:
Actually, as appeared - M0_IC_LIST lists the indexes, not keys. So you cannot use it.

So going back to your question - no, there is no API to list only the keys without values in Motr.